compliance/collection/main.go

```diff
@@ -182,17 +217,106 @@ func manageNodeScanLoop(ctx context.Context, rescanInterval time.Duration, scann
+func scanNodeWithBackoff(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	backoffIntervalSeconds := int64(env.NodeInventoryInitialBackoff.DurationSetting() / time.Second)
+
+	backoffFile, err := os.ReadFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath))
+	defer func() {
+		if err := os.Remove(fmt.Sprintf("%s/backoff", opts.inventoryCachePath)); err != nil {
+			log.Warnf("Could not remove backoff state file: %v", err)
+		}
+	}()
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No backoff found, continuing without pause")
+		} else {
+			return nil, err
+		}
+	} else {
+		// We have an existing backoff counter
+		backoffIntervalMillis, err := strconv.ParseInt(string(backoffFile), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		backoffIntervalSeconds = int64(time.Duration(backoffIntervalMillis) / time.Second)
+		log.Infof("Found existing backoff. Waiting before running next inventory")
+		backoffIntervalSeconds = opts.backoffWaitCallback(backoffIntervalSeconds)
+	}
+
+	err = os.WriteFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath), []byte(fmt.Sprintf("%d", backoffIntervalSeconds)), 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	message, err := cachedScanNode(opts)
+	if err != nil {
+		return nil, err
+	}
+	return message, nil
+}
+
+// cachedScanNode checks for a cached inventory before running a new scan
+func cachedScanNode(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	var inventory *storage.NodeInventory
+
+	cachedInv, err := os.ReadFile(fmt.Sprintf("%s/last_scan", opts.inventoryCachePath))
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No cache file found, running new inventory")
+		}
+		// Running into any other error than ErrNotExist continues to a fresh inventory
+		log.Infof("Unable to read inventory cache, running new inventory. Error: %v", err)
+	} else {
+		// try reading the inventory cached from disk and check whether it is new enough to send
+		cachedInvMsg := &storage.NodeInventory{}
+		if e := proto.Unmarshal(cachedInv, cachedInvMsg); e != nil {
+			// in this case, also collect a fresh inventory
+			log.Infof("Unable to deserialize inventory cache - running new inventory. Error: %v", e)
+		} else {
+			scanTime := cachedInvMsg.GetScanTime()
+			cacheThreshold := timestamp.TimestampNow().GetSeconds() - int64(env.NodeInventoryCacheDuration.DurationSetting().Seconds())
+			if scanTime != nil && scanTime.GetSeconds() > cacheThreshold {
+				log.Debugf("Using cached scan from %v", cachedInvMsg.GetScanTime())
+				return createAndObserveMessage(cachedInvMsg.GetNodeName(), cachedInvMsg), nil
+			}
+			log.Debugf("Cached scan older than threshold of %v with timestamp %v - running new inventory", env.NodeInventoryCacheDuration.DurationSetting(), cachedInvMsg.GetScanTime())
+		}
+	}
+
+	// Collect a fresh inventory
+	inventory, err = runCachedScan(opts)
```

Comment:
Something's definitely up with the function names. `cachedScanNode` is calling `runCachedScan`. Which one does what I can't tell from the function name.
Moreover, the way functions are composed forms rather deep call stack that (again) mixes multiple concerns.
Maybe the algorithm can be flattened and expressed in one high-level function that delegates boring tasks to its subordinate functions?
The algorithm in the high-level function would look something like this:

```
1. Read stored backoff and figure the current delay
2. Wait for the backoff
3. Store new backoff for subsequent runs
4. Read, deserialize cache and figure whether it's still usable
5. If cache is usable, use that for the resulting value
6. If cache is unusable or absent:
6.1. Perform the actual scan, use that for the resulting value
6.2. If the scan succeeds, cache it
7. Delete the backoff
8. Return the resulting value
```

Reply:
I have restructured the code into its own file in https://github.com/stackrox/stackrox/pull/4701/commits/3e6f0f9e9ba1b47a02ac1eb89b6ba359da80b5e7. This also included a rework revolving durations, testing, and the main function